chore: backport roughly a year's worth of changes

davidbarsky opened the pull request:
Bunch of backports, including:
- https://github.com/tokio-rs/tracing/pull/2373
- https://github.com/tokio-rs/tracing/pull/2377
- https://github.com/tokio-rs/tracing/pull/2362
- https://github.com/tokio-rs/tracing/pull/2426
- https://github.com/tokio-rs/tracing/pull/2446
- https://github.com/tokio-rs/tracing/pull/2389
- https://github.com/tokio-rs/tracing/pull/2647
- https://github.com/tokio-rs/tracing/pull/2508
- https://github.com/tokio-rs/tracing/pull/2708
- https://github.com/tokio-rs/tracing/pull/2617
- https://github.com/tokio-rs/tracing/pull/2690

hawkw commented:
I'm pretty sure some of the commits in here were not backported because they weren't relevant to the v0.1.x branch. For example, https://github.com/tokio-rs/tracing/pull/2728/commits/394207a108744cbb1614be3c6b1d42e6e4384c3e relates to the custom CSS we add to the rustdoc on v0.2.x to indicate that it's not the release version of `tracing`, which doesn't need to be on v0.1.x.

hawkw commented:
I'm also not really sure if https://github.com/tokio-rs/tracing/pull/2728/commits/9d98bb51875b07d68e44456338662db8dbf27595 needed to be backported: it reverted a commit that we never merged to v0.1.x, so now it just changes some docs link references.

At a glance, those are the only commits that I think didn't need to be backported.

davidbarsky commented:
Thanks, I'll remove those commits from this branch. https://github.com/davidbarsky/tracing-patcher might've been a little too aggressive.

davidbarsky commented:
Quick updates:
- I've removed 394207a and 9d98bb5 from this branch.
- fixed the `tracing-journald` compilation issues (didn't catch this because it was a clean rebase _and_ I'm on a Mac, so it didn't compile in the first place).
- Fixed 8372ba3 to ensure that its tests reference `fmt::Layer` instead of `fmt::Subscriber`, which has a _very_ different meaning on v0.1.x. 
    - I also changed 8372ba3's commit message to reference `fmt::Layer` instead of `fmt::Subscriber` to make it a little less confusing, but I can undo this change if it's a little *too* fishy.
- I've missed a necessary change in f23de16 while rebasing, but I'll fix that after I post this comment.

davidbarsky commented:
CI is currently failing due to [`tracing/tests/macros_redefined_core.rs`](https://github.com/tokio-rs/tracing/blob/430206459badca1f52963f3a6349784a5cea42aa/tracing/tests/macros_redefined_core.rs)'s usage of `extern crate self as core;`. I've made a [minimal reproduction](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f7f049b42617658eac970c87fb2d1a96).

davidbarsky commented:
@hawkw: I'm not sure when `tracing/tests/macros_redefined_core.rs` started causing issues, as building against different commits in this this branches's history still caused that compilation issue. To unblock/surface any other issues, I'm going to remove the test for now and try to figure it out later.

@steveklabnik (feel free to unsubscribe yourself, Steve! Just trying to give credit) suggested that it might be a lookup bug in `rustc`, noting that `extern crate self as core` is deeply fucked up and the compiler assumes that nobody is gonna add a crate named `core` without _also_ doing `no_core`. However, I compiled my branch with 1.56.0 (with some version patching, of course) and still encountered that compilation issue, so I'm not sure it was a _recent_ regression.

steveklabnik commented:
Just to be 1000% clear that was a guess of mine, I don't actually know if the compiler makes assumptions around core, just that it wouldn't shock me if it did.

davidbarsky commented:
CI now passes modulo two details:
1. I removed a test that messed with macro expansion in 0fc4626. I _think_ I we should try and keep this test, but note this test isn't on master and I don't have the faintest idea as to why it started failing.
2. I had to increase the minimum version of `proc-macro2` in `tracing-attributes` to [1.0.60](https://github.com/dtolnay/proc-macro2/releases/tag/1.0.60) from 1.0.40 from  to work around a failure in the `-Zminimal-versions` check. I'll make this change in a separate PR and rebase this branch against it.

davidbarsky commented:
I've isolated #2621 as the cause for `tracing/tests/macros_redefined_core.rs` failing to compile, most noisily under `cargo nextest run --profile ci --workspace`.

hawkw commented on tracing-subscriber/tests/layer_filters/main.rs:
```diff
@@ -2,33 +2,34 @@
 
 #[test]
-fn basic_layer_filters() {
+fn basic_subscriber_filters() {
```
can we change this to use the v0.1.x names? e.g. `basic_layer_filters`, as it was previously?

hawkw commented on tracing-subscriber/tests/layer_filters/main.rs:
```diff
@@ -47,23 +48,23 @@ fn basic_layer_filters() {
 
 #[test]
-fn basic_layer_filters_spans() {
+fn basic_subscriber_filters_spans() {
```
can we change this to use the v0.1.x names? e.g. `basic_subscriber_filters_spans`, as it was previously?

hawkw commented on tracing-subscriber/tests/layer_filters/option.rs:
```diff
@@ -0,0 +1,149 @@
+    let subscribe_none = layer_some.with_filter(None::<filter::DynFilterFn<_>>);
+    assert!(subscribe_none.max_level_hint().is_none());
+
+    let (subscribe_filter_fn, handle_filter_fn) = layer::mock()
```
can we change this to use the v0.1.x naming? e.g. `layer_filter_fn`?

hawkw commented on tracing-subscriber/tests/layer_filters/option.rs:
```diff
@@ -0,0 +1,149 @@
+        .only()
+        .run_with_handle();
+    let max_level = Level::INFO;
+    let subscribe_filter_fn = subscribe_filter_fn.with_filter(
+        filter::dynamic_filter_fn(move |meta, _| return meta.level() <= &max_level)
+            .with_max_level_hint(max_level),
+    );
+    assert_eq!(
+        subscribe_filter_fn.max_level_hint(),
+        Some(LevelFilter::INFO)
+    );
+
+    let subscriber = tracing_subscriber::registry()
+        .with(subscribe_none)
+        .with(subscribe_filter_fn);
+    // The absence of a hint from the `None` filter upgrades the `INFO` hint
+    // from the filter fn subscriber.
+    assert!(subscriber.max_level_hint().is_none());
```
more uses of v0.2 names (subscriber rather than layer)

davidbarsky replied:
I'm going to keep the naming scheme of the `let subscriber`... starting on line 93 _as_ a subscriber, not a layer. I'm pretty confident that those are your intentions.

hawkw commented on tracing-subscriber/tests/layer_filters/option.rs:
```diff
@@ -0,0 +1,149 @@
+    ));
+    assert_eq!(layer_some.max_level_hint(), Some(LevelFilter::DEBUG));
+
+    let (subscribe_filter_fn, handle_filter_fn) = layer::mock()
+        .event(expect::event())
+        .only()
+        .run_with_handle();
+    let subscribe_filter_fn = subscribe_filter_fn.with_filter(
+        filter::dynamic_filter_fn(move |meta, _| return meta.level() <= &Level::INFO)
+            .with_max_level_hint(Level::INFO),
+    );
+    assert_eq!(
+        subscribe_filter_fn.max_level_hint(),
+        Some(LevelFilter::INFO)
+    );
+
+    let subscriber = tracing_subscriber::registry()
+        .with(layer_some)
+        .with(subscribe_filter_fn);
+    // The `DEBUG` hint from the `Some` filter upgrades the `INFO` hint from the
+    // filter fn subscriber.
+    assert_eq!(subscriber.max_level_hint(), Some(LevelFilter::DEBUG));
```
wrong naming here as well

hawkw commented on tracing-subscriber/tests/vec_subscriber_filter_interests_cached.rs:
```diff
@@ -4,14 +4,11 @@ use std::{
 
 #[test]
-fn vec_layer_filter_interests_are_cached() {
+fn vec_subscriber_filter_interests_are_cached() {
```
let's keep the v0.1.x naming here please (`vec_layer_filter_interests_are_cached`)

hawkw commented on tracing/tests/no_subscriber.rs:
```diff
@@ -1,14 +1,15 @@
 #[cfg_attr(target_arch = "wasm32", wasm_bindgen_test::wasm_bindgen_test)]
 #[test]
-fn no_subscriber_disables_global() {
+fn no_collector_disables_global() {
```
let's use the v0.1.x naming for this please (`no_subscriber...` rather than `no_collector...`)

davidbarsky replied:
whoops, missed this, sorry.

hawkw reviewed (commented):
overall, this looks good. i noticed some accidental uses of the `tracing` v0.2 type names in the tests, can we change those back to the v0.1.x names?

other than that, this looks great to me!

davidbarsky commented on tracing-subscriber/src/filter/env/directive.rs:
```diff
@@ -120,8 +120,9 @@ impl Directive {
-            r"(?x)
+        static DIRECTIVE_RE: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?x)
```
i think it was rustfmt, yeah.

hawkw replied:
okay, cool, that's fine. super weird but if rustfmt likes it, whatever!

davidbarsky commented:
I've noticed a few more mentions of 0.2 naming after pushing, sorry. I'll fix those up.

davidbarsky commented:
> I've noticed a few more mentions of 0.2 naming after pushing, sorry. I'll fix those up.

Alright, I think I got rid of any 0.2 terminology.

hawkw reviewed (approved):
looks great, thank you so much!
